src/Control/Monad/Free/Freer.hs

```diff
@@ -65,13 +75,23 @@ iterA algebra = iterFreerA ((algebra .) . flip fmap)
 iterFreer :: (forall x. f x -> (x -> a) -> a) -> Freer f a -> a
 iterFreer algebra = cata $ \ r -> case r of
   ReturnF result -> result
+  MapF transform action -> algebra action transform
   ThenF action continue -> algebra action continue
 {-# INLINE iterFreer #-}
 
 iterFreerA :: Applicative m => (forall x. f x -> (x -> m a) -> m a) -> Freer f a -> m a
 iterFreerA algebra r = iterFreer algebra (fmap pure r)
 {-# INLINE iterFreerA #-}
 
+iterLookahead :: forall f a. (forall x y. f x -> Maybe (Freer f y) -> (x -> a) -> a) -> Freer f a -> a
+iterLookahead algebra = go
+  where go :: Freer f a -> a
+        go (Return a) = a
+        go (Map f a) = algebra a Nothing f
+        go (Seq f a b) = algebra a (Just b) (go . flip fmap b . f)
+        go (Then a f) = algebra a Nothing (go . f)
+{-# INLINE iterLookahead #-}
```

Comment:
I’m not sure just yet whether this will cut it, but `iterLookahead` provides the algebra with the next step, when it’s available. In theory, this should accommodate algorithms requiring lookahead.